helm/config.go

```diff
@@ -0,0 +1,377 @@
+func (enc *Encoder) Set(modifiers ...func(*Encoder)) {
+	for _, modifier := range modifiers {
+		modifier(enc)
+	}
+}
+
+// NewEncoder returns an Encoder object holding the output stream and encoding options.
+func NewEncoder(writer io.Writer, modifiers ...func(*Encoder)) *Encoder {
+	enc := &Encoder{
+		// Wrap io.Writer in a bufio.Writer so that we can check for
+		// errors at the very end by calling bufio.Writer.Flush()
+		writer:     bufio.NewWriter(writer),
+		emptyLines: false,
+		indent:     2,
+		wrap:       80,
+	}
+	enc.Set(modifiers...)
+	return enc
+}
+
+// Encode writes the config object to the stream.
+func (enc *Encoder) Encode(object *Object) error {
+	enc.pendingNewline = false
+	fmt.Fprintln(enc.writer, "---")
+	prefix := ""
+	enc.writeNode(object, &prefix, "")
+	return enc.writer.(*bufio.Writer).Flush()
+}
+
+// useOnce returns the current value of the prefix parameter but replaces it
+// with a string of spaces of the same length as the original prefix for
+// subsequent use. This is done because for nested list elements the "- "
+// prefixes are being stacked, and should only be used in front of the first
+// element of the innermost list:
+//
+//   [ [ [ 1, 2 ] ] ]   --->   " - - - 1\n    - 2\n"
+//
+func useOnce(prefix *string) string {
+	result := *prefix
+	*prefix = strings.Repeat(" ", len(*prefix))
+	return result
+}
+
+// writeComment writes out the comment block for a node. Newline characters in
+// the comment mark the beginning of a new paragraph. Each paragraph is
+// word-wrapped to fit within enc.wrap columns (except that each line will
+// include at least a single word, even if it exceeds the wrapping column).
+func (enc *Encoder) writeComment(prefix *string, comment string) {
+	for _, line := range strings.Split(comment, "\n") {
+		fmt.Fprintf(enc.writer, "%s#", useOnce(prefix))
+		if len(line) > 0 {
+			written := 0
+			for _, word := range strings.Fields(line) {
+				if written > 0 && len(*prefix)+1+written+1+len(word) > enc.wrap {
+					fmt.Fprintf(enc.writer, "\n%s#", useOnce(prefix))
+					written = 0
+				}
+				fmt.Fprint(enc.writer, " "+word)
+				written += 1 + len(word)
+			}
+		}
+		fmt.Fprint(enc.writer, "\n")
+	}
+}
+
+// writeNode is called for each element of a container (list or object node). It
+// will print the comment and condition for the element and then call
+// node.write() to tell the node to encode itself.
+//
+// prefix includes indentation and the label for the container of this element
+// (either "Name:" or a string of one or more "- "). It will only be printed in
+// front of the first element of the container and then be replaced with a
+// string of spaces for indentation of all latter line. The label is not printed
+// by the container itself because the list labels "- " stack up on a single line.
+//
+// label contains either the "Name:" for object elements, or "-" for list
+// elements.
+func (enc *Encoder) writeNode(node Node, prefix *string, label string) {
```

Comment:
Finding this method a bit hard to go through, but I have no idea how to make it clearer.